File: Test Case Generation/deepseek/cpp/test_task_7.cpp
#include <gtest/gtest.h>
#include <vector>
#include <string>
#include "filter_by_substring.h"  // Assumes the function is declared here

using namespace std;

// Test fixture for filter_by_substring tests
class FilterBySubstringTest : public ::testing::Test {
protected:
    void SetUp() override {}
    void TearDown() override {}
};

// Test 1: Empty input vector
TEST_F(FilterBySubstringTest, EmptyVector) {
    vector<string> input;
    string substring = "a";
    vector<string> result = filter_by_substring(input, substring);
    EXPECT_TRUE(result.empty());
    EXPECT_EQ(0, result.size());
}

// Test 2: Empty substring
TEST_F(FilterBySubstringTest, EmptySubstring) {
    vector<string> input = {"abc", "def", "abc def"};
    string substring = "";
    vector<string> result = filter_by_substring(input, substring);
    
    // Empty substring is found in every string
    vector<string> expected = {"abc", "def", "abc def"};
    EXPECT_EQ(expected, result);
}

// Test 3: Substring present in multiple elements
TEST_F(FilterBySubstringTest, BasicFiltering) {
    vector<string> input = {"abc", "bacd", "cde", "vector"};
    string substring = "a";
    vector<string> result = filter_by_substring(input, substring);
    
    vector<string> expected = {"abc", "bacd", "vector"};
    EXPECT_EQ(expected, result);
}

// Test 4: No matches
TEST_F(FilterBySubstringTest, NoMatches) {
    vector<string> input = {"xyz", "123", "def"};
    string substring = "a";
    vector<string> result = filter_by_substring(input, substring);
    
    EXPECT_TRUE(result.empty());
}

// Test 5: Substring at beginning
TEST_F(FilterBySubstringTest, SubstringAtBeginning) {
    vector<string> input = {"apple", "apricot", "banana", "grape"};
    string substring = "ap";
    vector<string> result = filter_by_substring(input, substring);
    
    vector<string> expected = {"apple", "apricot"};
    EXPECT_EQ(expected, result);
}

// Test 6: Substring at end
TEST_F(FilterBySubstringTest, SubstringAtEnd) {
    vector<string> input = {"running", "jumping", "swimming", "walking"};
    string substring = "ing";
    vector<string> result = filter_by_substring(input, substring);
    
    vector<string> expected = {"running", "jumping", "swimming", "walking"};
    EXPECT_EQ(expected, result);
}

// Test 7: Substring in middle
TEST_F(FilterBySubstringTest, SubstringInMiddle) {
    vector<string> input = {"hello world", "good morning", "happy birthday", "test"};
    string substring = "o";
    vector<string> result = filter_by_substring(input, substring);
    
    vector<string> expected = {"hello world", "good morning"};
    EXPECT_EQ(expected, result);
}

// Test 8: Single element vector with match
TEST_F(FilterBySubstringTest, SingleElementMatch) {
    vector<string> input = {"hello"};
    string substring = "ell";
    vector<string> result = filter_by_substring(input, substring);
    
    vector<string> expected = {"hello"};
    EXPECT_EQ(expected, result);
}

// Test 9: Single element vector without match
TEST_F(FilterBySubstringTest, SingleElementNoMatch) {
    vector<string> input = {"hello"};
    string substring = "xyz";
    vector<string> result = filter_by_substring(input, substring);
    
    EXPECT_TRUE(result.empty());
}

// Test 10: Empty strings in vector
TEST_F(FilterBySubstringTest, EmptyStringsInVector) {
    vector<string> input = {"", "abc", "", "def", ""};
    string substring = "abc";
    vector<string> result = filter_by_substring(input, substring);
    
    vector<string> expected = {"abc"};
    EXPECT_EQ(expected, result);
}

// Test 11: Empty string in vector with empty substring
TEST_F(FilterBySubstringTest, EmptyStringVectorWithEmptySubstring) {
    vector<string> input = {"", "abc", ""};
    string substring = "";
    vector<string> result = filter_by_substring(input, substring);
    
    vector<string> expected = {"", "abc", ""};
    EXPECT_EQ(expected, result);
}

// Test 12: Substring with special characters
TEST_F(FilterBySubstringTest, SpecialCharacters) {
    vector<string> input = {"hello@world", "test#123", "hello#world", "normal"};
    string substring = "#";
    vector<string> result = filter_by_substring(input, substring);
    
    vector<string> expected = {"test#123", "hello#world"};
    EXPECT_EQ(expected, result);
}

// Test 13: Substring with spaces
TEST_F(FilterBySubstringTest, SubstringWithSpaces) {
    vector<string> input = {"hello world", "helloworld", "hello  world", "hi there"};
    string substring = "lo w";
    vector<string> result = filter_by_substring(input, substring);
    
    vector<string> expected = {"hello world"};
    EXPECT_EQ(expected, result);
}

// Test 14: Large vector
TEST_F(FilterBySubstringTest, LargeVector) {
    vector<string> input;
    for (int i = 0; i < 1000; ++i) {
        input.push_back("string_" + to_string(i));
    }
    
    string substring = "500";
    vector<string> result = filter_by_substring(input, substring);
    
    EXPECT_EQ(1, result.size());
    EXPECT_EQ("string_500", result[0]);
}

// Test 15: Substring appears multiple times in same string
TEST_F(FilterBySubstringTest, MultipleOccurrences) {
    vector<string> input = {"aaaa", "aaab", "baaa", "abc"};
    string substring = "aa";
    vector<string> result = filter_by_substring(input, substring);
    
    vector<string> expected = {"aaaa", "aaab", "baaa"};
    EXPECT_EQ(expected, result);
}

// Test 16: Case sensitivity
TEST_F(FilterBySubstringTest, CaseSensitivity) {
    vector<string> input = {"Hello", "hello", "HELLO", "HeLLo"};
    string substring = "Hell";
    vector<string> result = filter_by_substring(input, substring);
    
    // Only exact match with case sensitivity
    vector<string> expected = {"Hello"};
    EXPECT_EQ(expected, result);
}

// Test 17: Single character substring
TEST_F(FilterBySubstringTest, SingleCharacter) {
    vector<string> input = {"abc", "def", "ghi", "jkl"};
    string substring = "d";
    vector<string> result = filter_by_substring(input, substring);
    
    vector<string> expected = {"def"};
    EXPECT_EQ(expected, result);
}

// Test 18: Very long substring
TEST_F(FilterBySubstringTest, LongSubstring) {
    vector<string> input = {"short", "this is a very long string for testing", "medium"};
    string substring = "very long string for testing";
    vector<string> result = filter_by_substring(input, substring);
    
    vector<string> expected = {"this is a very long string for testing"};
    EXPECT_EQ(expected, result);
}

// Test 19: Substring longer than any string
TEST_F(FilterBySubstringTest, SubstringLongerThanString) {
    vector<string> input = {"short", "tiny", "small"};
    string substring = "verylongsubstring";
    vector<string> result = filter_by_substring(input, substring);
    
    EXPECT_TRUE(result.empty());
}

// Test 20: Verify no modification of original vector
TEST_F(FilterBySubstringTest, OriginalVectorUnmodified) {
    vector<string> input = {"abc", "def", "ghi"};
    vector<string> original = input;
    string substring = "d";
    
    filter_by_substring(input, substring);
    
    EXPECT_EQ(original, input);
}

// Main function for running tests
int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: Test Case Generation/deepseek/python/test_task_7.py
import pytest
from typing import List
from your_module import filter_by_substring

def test_empty_list():
    """Test with empty list of strings"""
    assert filter_by_substring([], 'a') == []

def test_empty_substring():
    """Test with empty substring (should match all strings)"""
    strings = ['abc', 'def', 'ghi']
    assert filter_by_substring(strings, '') == ['abc', 'def', 'ghi']

def test_single_match():
    """Test with single matching string"""
    strings = ['abc', 'def', 'ghi']
    assert filter_by_substring(strings, 'ab') == ['abc']

def test_multiple_matches():
    """Test with multiple matching strings"""
    strings = ['abc', 'bacd', 'cde', 'array']
    assert filter_by_substring(strings, 'a') == ['abc', 'bacd', 'array']

def test_no_matches():
    """Test with no matching strings"""
    strings = ['abc', 'def', 'ghi']
    assert filter_by_substring(strings, 'z') == []

def test_all_match():
    """Test where all strings contain the substring"""
    strings = ['cat', 'bat', 'hat']
    assert filter_by_substring(strings, 'at') == ['cat', 'bat', 'hat']

def test_substring_at_beginning():
    """Test substring at the beginning of strings"""
    strings = ['prefix1', 'prefix2', 'suffix']
    assert filter_by_substring(strings, 'pre') == ['prefix1', 'prefix2']

def test_substring_at_end():
    """Test substring at the end of strings"""
    strings = ['hello.py', 'world.py', 'test.txt']
    assert filter_by_substring(strings, '.py') == ['hello.py', 'world.py']

def test_substring_in_middle():
    """Test substring in the middle of strings"""
    strings = ['startmiddleend', 'beginmiddlefinish', 'nomatch']
    assert filter_by_substring(strings, 'middle') == ['startmiddleend', 'beginmiddlefinish']

def test_case_sensitivity():
    """Test case-sensitive matching"""
    strings = ['Apple', 'apple', 'APPLE']
    assert filter_by_substring(strings, 'App') == ['Apple']
    assert filter_by_substring(strings, 'app') == ['apple']
    assert filter_by_substring(strings, 'APP') == ['APPLE']

def test_special_characters():
    """Test with special characters in strings and substring"""
    strings = ['hello!@#', 'world$%^', 'test&*(']
    assert filter_by_substring(strings, '!@#') == ['hello!@#']
    assert filter_by_substring(strings, '^&') == []

def test_unicode_characters():
    """Test with unicode characters"""
    strings = ['café', 'école', 'français']
    assert filter_by_substring(strings, 'é') == ['café', 'école', 'français']

def test_numbers():
    """Test with numeric strings and substrings"""
    strings = ['123', '456', '7890']
    assert filter_by_substring(strings, '3') == ['123']
    assert filter_by_substring(strings, '90') == ['7890']

def test_whitespace():
    """Test with whitespace characters"""
    strings = ['hello world', 'hello\tworld', 'helloworld']
    assert filter_by_substring(strings, ' ') == ['hello world']
    assert filter_by_substring(strings, '\t') == ['hello\tworld']

def test_long_strings():
    """Test with very long strings"""
    long_string = 'a' * 1000 + 'target' + 'b' * 1000
    strings = [long_string, 'no match', 'target']
    assert filter_by_substring(strings, 'target') == [long_string, 'target']

def test_duplicate_matches():
    """Test with duplicate strings in input"""
    strings = ['abc', 'abc', 'def', 'abc']
    assert filter_by_substring(strings, 'ab') == ['abc', 'abc', 'abc']

def test_overlapping_substring():
    """Test substring appears in overlapping pattern"""
    strings = ['aaaaaa']
    assert filter_by_substring(strings, 'aa') == ['aaaaaa']

def test_none_list():
    """Test with None as list input should raise TypeError"""
    with pytest.raises(TypeError):
        filter_by_substring(None, 'a')

def test_none_substring():
    """Test with None as substring should raise TypeError"""
    with pytest.raises(TypeError):
        filter_by_substring(['abc', 'def'], None)

def test_non_string_elements():
    """Test with list containing non-string elements"""
    with pytest.raises(TypeError):
        filter_by_substring(['abc', 123, 'def'], 'a')

def test_empty_string_element():
    """Test with empty string in list"""
    strings = ['', 'abc', '', 'def']
    assert filter_by_substring(strings, '') == ['', 'abc', '', 'def']
    assert filter_by_substring(strings, 'a') == ['abc']

def test_single_character_substring():
    """Test with single character substring"""
    strings = ['a', 'b', 'c', 'ab', 'bc']
    assert filter_by_substring(strings, 'a') == ['a', 'ab']

def test_substring_longer_than_string():
    """Test where substring is longer than any string"""
    strings = ['short', 'tiny', 'small']
    assert filter_by_substring(strings, 'verylongsubstring') == []

def test_preserve_order():
    """Test that original order is preserved"""
    strings = ['zebra', 'apple', 'banana', 'grape']
    result = filter_by_substring(strings, 'a')
    assert result == ['apple', 'banana', 'grape']
    assert result.index('apple') < result.index('banana') < result.index('grape')

def test_return_new_list():
    """Test that function returns a new list, not modifying original"""
    original = ['abc', 'def', 'ghi']
    result = filter_by_substring(original, 'a')
    assert result is not original
    assert original == ['abc', 'def', 'ghi']